File: quickstart/sam.py
"""SAM lifecycle management: build, start local, wait, stop, deploy."""

from __future__ import annotations

import subprocess
import time
import urllib.error
import urllib.request
from pathlib import Path

from quickstart import runner
from quickstart.config import ProjectConfig

# Lambda env var overrides for sam local (no DynamoDB available locally)
_LOCAL_ENV_OVERRIDES = {
    "HelloWorldFunction": {
        "STORE_BACKEND": "sqlite",
    }
}


class SamLocal:
    """Context manager for sam local start-api lifecycle."""

    def __init__(
        self,
        config: ProjectConfig,
        *,
        env: dict[str, str] | None = None,
        verbose: bool = False,
        dry_run: bool = False,
    ):
        self._config = config
        self._env = env
        self._verbose = verbose
        self._dry_run = dry_run
        self._proc: subprocess.Popen | None = None

    @property
    def url(self) -> str:
        return f"http://localhost:{self._config.sam_local_port}"

    def build(self, *, no_cache: bool = False) -> None:
        """Export runtime deps and run sam build."""
        log_dir = self._config.log_dir

        runner.step("Exporting runtime dependencies")
        runner.run(
            ["uv", "export", "--no-dev", "--no-hashes", "-o", "requirements.txt"],
            cwd=self._config.cloud_dir / "uplink",
            env=self._env,
            verbose=self._verbose,
            dry_run=self._dry_run,
            log_to=log_dir / "uv_export.log",
        )

        runner.step("Building SAM artifacts")
        cmd = ["sam", "build", "--skip-pull-image"]
        if no_cache:
            cmd.append("--no-cached")
        runner.run(
            cmd,
            cwd=self._config.cloud_dir,
            env=self._env,
            verbose=self._verbose,
            dry_run=self._dry_run,
            log_to=log_dir / "sam_build.log",
        )
        runner.success("SAM build complete")

    def _write_env_overrides(self) -> Path:
        """Write Lambda env var overrides to a temp JSON file for --env-vars."""
        import json
        import tempfile

        env_file = Path(tempfile.mktemp(suffix=".json", prefix="sam_env_"))
        env_file.write_text(json.dumps(_LOCAL_ENV_OVERRIDES))
        return env_file

    def start(self, *, extra_args: list[str] | None = None) -> None:
        """Start sam local start-api in background."""
        runner.step(f"Starting sam local on port {self._config.sam_local_port}")
        self._env_file = self._write_env_overrides()
        cmd = [
            "sam",
            "local",
            "start-api",
            "--port",
            str(self._config.sam_local_port),
            "--skip-pull-image",
            "--env-vars",
            str(self._env_file),
        ]
        if extra_args:
            cmd.extend(extra_args)
        self._proc = runner.start_background(
            cmd,
            cwd=self._config.cloud_dir,
            env=self._env,
            log_file=self._config.sam_log_file,
            verbose=self._verbose,
            dry_run=self._dry_run,
        )

    def wait_ready(self) -> None:
        """Poll until sam local's HTTP server is up (any HTTP response)."""
        if self._dry_run:
            print("  [dry-run] wait for sam local ready")
            return

        # Hit a non-existent path — SAM responds immediately with 403/404
        # without invoking a Lambda, so we don't pay the cold-start penalty here.
        probe_url = f"{self.url}/_qs_health_probe"
        print(f"  Waiting for sam local at {self.url} ...")
        deadline = time.monotonic() + self._config.sam_ready_timeout

        while time.monotonic() < deadline:
            try:
                resp = urllib.request.urlopen(probe_url, timeout=2)
                runner.success(f"  sam local ready (HTTP {resp.status}).")
                return
            except urllib.error.HTTPError as e:
                # 4xx still means the HTTP server is up
                runner.success(f"  sam local ready (HTTP {e.code}).")
                return
            except (urllib.error.URLError, OSError):
                time.sleep(1)

        raise TimeoutError(
            f"sam local did not start within {self._config.sam_ready_timeout}s. "
            f"Check logs: {self._config.sam_log_file}"
        )

    def stop(self) -> None:
        """Terminate sam local process and clean up temp files."""
        if self._proc is not None and self._proc.poll() is None:
            print("  Stopping sam local...")
            self._proc.terminate()
            try:
                self._proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._proc.kill()
                self._proc.wait()
        if hasattr(self, "_env_file") and self._env_file.exists():
            self._env_file.unlink()

    def _read_stack_name(self, config_env: str) -> str:
        """Read stack_name from env vars, falling back to samconfig.toml."""
        # CLI flags from .env files take precedence (see deploy())
        env = self._env or {}
        if env.get("SAM_STACK_NAME"):
            return env["SAM_STACK_NAME"]

        import tomllib

        samconfig = self._config.cloud_dir / "samconfig.toml"
        with open(samconfig, "rb") as f:
            data = tomllib.load(f)

        env_section = data.get(config_env, {})
        # Try deploy.parameters first, then global.parameters
        for path in (("deploy", "parameters"), ("global", "parameters")):
            section = env_section
            for key in path:
                section = section.get(key, {})
            name = section.get("stack_name")
            if name:
                return name.strip('"')

        raise RuntimeError(
            f"No stack_name found in samconfig.toml for config-env '{config_env}'"
        )

    def stack_endpoint(self, config_env: str) -> str:
        """Query CloudFormation for the deployed API endpoint URL."""
        if self._dry_run:
            return f"https://DRY-RUN.execute-api.us-east-1.amazonaws.com/{config_env}"

        stack_name = self._read_stack_name(config_env)
        result = runner.run(
            [
                "aws", "cloudformation", "describe-stacks",
                "--stack-name", stack_name,
                "--query",
                "Stacks[0].Outputs[?OutputKey=='SupervictorApiEndpoint'].OutputValue",
                "--output", "text",
            ],
            env=self._env,
            verbose=self._verbose,
            dry_run=self._dry_run,
            capture=True,
        )
        url = result.stdout.strip().rstrip("/")
        if not url:
            raise RuntimeError(
                f"No SupervictorApiEndpoint output found for stack '{stack_name}'"
            )
        return url

    def deploy(self, config_env: str, *, force_upload: bool = False) -> bool:
        """Run sam deploy --config-env <env>. Returns True if changes deployed.

        If the process env (self._env) contains SAM_* variables, they are passed
        as CLI flags so that .env.dev / .env.staging drives the deploy instead of
        hardcoded values in samconfig.toml.
        """
        log_path = self._config.log_dir / f"sam_deploy_{config_env}.log"
        runner.step(f"Deploying to {config_env} stack")
        cmd = ["sam", "deploy", "--config-env", config_env]

        # Override samconfig values from env vars when present
        env = self._env or {}
        if env.get("SAM_STACK_NAME"):
            cmd.extend(["--stack-name", env["SAM_STACK_NAME"]])
        if env.get("SAM_REGION"):
            cmd.extend(["--region", env["SAM_REGION"]])
        if env.get("SAM_S3_PREFIX"):
            cmd.extend(["--s3-prefix", env["SAM_S3_PREFIX"]])

        # Build --parameter-overrides from SAM_* env vars
        param_parts: list[str] = []
        param_map = {
            "SAM_ENVIRONMENT": "Environment",
            "SAM_APP_NAME": "AppName",
            "SAM_STACK_NAME": "StackName",
            "SAM_TRUSTSTORE_URI": "TruststoreUri",
        }
        for env_key, cfn_param in param_map.items():
            val = env.get(env_key)
            if val:
                param_parts.append(f"{cfn_param}={val}")
        if param_parts:
            cmd.extend(["--parameter-overrides", " ".join(param_parts)])

        if force_upload:
            cmd.append("--force-upload")
        result = runner.run(
            cmd,
            cwd=self._config.cloud_dir,
            env=self._env,
            verbose=self._verbose,
            dry_run=self._dry_run,
            check=False,
            log_to=log_path,
        )
        if result.returncode != 0:
            if "No changes to deploy" in (result.stderr or "") + (result.stdout or ""):
                runner.success("Stack is already up to date.")
                return False
            else:
                runner.error(f"sam deploy failed (see {log_path})")
                raise subprocess.CalledProcessError(result.returncode, result.args)
        runner.success(f"Deployed to {config_env}")
        return True

    def __enter__(self) -> SamLocal:
        self.start()
        self.wait_ready()
        return self

    def __exit__(self, *exc: object) -> None:
        self.stop()

File: cli/src/sam.rs
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::process::Child;
use std::time::{Duration, Instant};

use crate::config::ProjectConfig;
use crate::error::CliError;
use crate::runner::{self, BackgroundOptions, RunOptions, Runner};

/// Lambda env var overrides for sam local (no DynamoDB available locally).
const LOCAL_ENV_OVERRIDES: &str = r#"{"HelloWorldFunction":{"STORE_BACKEND":"sqlite"}}"#;

/// SAM local lifecycle manager.
pub struct SamLocal<'a> {
    config: &'a ProjectConfig,
    env: Option<HashMap<String, String>>,
    verbose: bool,
    dry_run: bool,
    proc: Option<Child>,
    env_file: Option<PathBuf>,
}

/// Guard returned by `start()` that terminates the process on drop.
pub struct SamGuard {
    proc: Option<Child>,
    env_file: Option<PathBuf>,
}

impl Drop for SamGuard {
    fn drop(&mut self) {
        if let Some(ref mut child) = self.proc {
            let _ = child.kill();
            let _ = child.wait();
        }
        if let Some(ref path) = self.env_file {
            let _ = fs::remove_file(path);
        }
    }
}

impl SamGuard {
    /// Take ownership of the child process, preventing the guard from killing it on drop.
    pub fn take_process(&mut self) -> Option<Child> {
        self.proc.take()
    }

    /// Block until the process exits.
    pub fn wait(&mut self) -> Result<(), CliError> {
        if let Some(ref mut child) = self.proc {
            child.wait()?;
        }
        Ok(())
    }

    /// Detach the process so it keeps running after this guard is dropped.
    /// Writes the PID to `pid_file` for later cleanup.
    pub fn detach(mut self, pid_file: &std::path::Path) -> Result<u32, CliError> {
        let child = self
            .proc
            .take()
            .ok_or_else(|| CliError::Config("no SAM process to detach (dry-run?)".to_string()))?;
        let pid = child.id();

        if let Some(parent) = pid_file.parent() {
            std::fs::create_dir_all(parent)?;
        }
        std::fs::write(pid_file, pid.to_string())?;

        // Leak the Child so Drop doesn't kill it
        std::mem::forget(child);
        Ok(pid)
    }
}

/// Kill a previously-detached SAM process using its PID file.
/// Returns Ok(true) if the process was killed, Ok(false) if it wasn't running.
pub fn stop_from_pid_file(pid_file: &std::path::Path) -> Result<bool, CliError> {
    let contents = match fs::read_to_string(pid_file) {
        Ok(c) => c,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            return Ok(false);
        }
        Err(e) => return Err(e.into()),
    };

    let pid = contents.trim().to_string();
    if pid.is_empty() {
        let _ = fs::remove_file(pid_file);
        return Ok(false);
    }

    // Use kill(1) to send SIGTERM — no libc dependency needed
    let status = std::process::Command::new("kill")
        .arg(&pid)
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status();

    let killed = status.is_ok_and(|s| s.success());

    // Clean up PID file regardless
    let _ = fs::remove_file(pid_file);

    Ok(killed)
}

impl<'a> SamLocal<'a> {
    pub fn new(
        config: &'a ProjectConfig,
        env: Option<HashMap<String, String>>,
        verbose: bool,
        dry_run: bool,
    ) -> Self {
        Self {
            config,
            env,
            verbose,
            dry_run,
            proc: None,
            env_file: None,
        }
    }

    pub fn url(&self) -> String {
        format!("http://localhost:{}", self.config.sam_local_port)
    }

    /// Export runtime deps and run sam build.
    pub fn build(&self, r: &dyn Runner, no_cache: bool) -> Result<(), CliError> {
        let log_dir = &self.config.log_dir;

        runner::step("Exporting runtime dependencies");
        r.run(
            &[
                "uv",
                "export",
                "--no-dev",
                "--no-hashes",
                "-o",
                "requirements.txt",
            ],
            &RunOptions {
                cwd: Some(self.config.cloud_dir.join("uplink")),
                env: self.env.clone(),
                verbose: self.verbose,
                dry_run: self.dry_run,
                log_to: Some(log_dir.join("uv_export.log")),
                ..Default::default()
            },
        )?;

        runner::step("Building SAM artifacts");
        let mut cmd: Vec<&str> = vec!["sam", "build", "--skip-pull-image"];
        if no_cache {
            cmd.push("--no-cached");
        }
        r.run(
            &cmd,
            &RunOptions {
                cwd: Some(self.config.cloud_dir.clone()),
                env: self.env.clone(),
                verbose: self.verbose,
                dry_run: self.dry_run,
                log_to: Some(log_dir.join("sam_build.log")),
                ..Default::default()
            },
        )?;
        runner::success("SAM build complete");
        Ok(())
    }

    /// Start sam local start-api in background and wait for readiness.
    /// Returns a guard that stops the process on drop.
    pub fn start(&mut self, r: &dyn Runner, extra_args: &[&str]) -> Result<SamGuard, CliError> {
        runner::step(&format!(
            "Starting sam local on port {}",
            self.config.sam_local_port
        ));

        let env_file = self.write_env_overrides()?;
        let env_file_str = env_file.to_string_lossy().to_string();
        let port_str = self.config.sam_local_port.to_string();

        let mut cmd = vec![
            "sam",
            "local",
            "start-api",
            "--port",
            &port_str,
            "--skip-pull-image",
            "--env-vars",
            &env_file_str,
        ];
        cmd.extend_from_slice(extra_args);

        let child = r.start_background(
            &cmd,
            &BackgroundOptions {
                cwd: Some(self.config.cloud_dir.clone()),
                env: self.env.clone(),
                log_file: Some(PathBuf::from(&self.config.sam_log_file)),
                verbose: self.verbose,
                dry_run: self.dry_run,
            },
        )?;

        self.env_file = Some(env_file.clone());
        self.proc = child;

        Ok(SamGuard {
            proc: self.proc.take(),
            env_file: Some(env_file),
        })
    }

    /// Poll until sam local's HTTP server is up and responding.
    pub fn wait_ready(&self) -> Result<(), CliError> {
        if self.dry_run {
            println!("  [dry-run] wait for sam local ready");
            return Ok(());
        }

        let health_url = format!("{}/health", self.url());
        println!("  Waiting for sam local at {} ...", self.url());
        let deadline = Instant::now() + Duration::from_secs(self.config.sam_ready_timeout);

        while Instant::now() < deadline {
            // Full HTTP probe — TCP alone succeeds before Lambda container is ready
            // Accept any HTTP response (even 403) — we just need the server responding
            match ureq::get(&health_url).call() {
                Ok(_) | Err(ureq::Error::StatusCode(_)) => {
                    runner::success("  sam local ready.");
                    return Ok(());
                }
                Err(_) => {}
            }
            std::thread::sleep(Duration::from_secs(1));
        }

        Err(CliError::Timeout {
            message: format!(
                "sam local did not start within {}s. Check logs: {}",
                self.config.sam_ready_timeout,
                self.config.sam_log_file.display()
            ),
        })
    }

    /// Read stack_name from env vars, falling back to samconfig.toml.
    fn read_stack_name(&self, config_env: &str) -> Result<String, CliError> {
        let env = self.env.as_ref();
        if let Some(name) = env.and_then(|e| e.get("SAM_STACK_NAME")) {
            if !name.is_empty() {
                return Ok(name.clone());
            }
        }

        let samconfig = self.config.cloud_dir.join("samconfig.toml");
        let contents = fs::read_to_string(&samconfig).map_err(|e| {
            CliError::Config(format!("failed to read {}: {}", samconfig.display(), e))
        })?;
        let data: toml::Value = contents
            .parse()
            .map_err(|e| CliError::Config(format!("failed to parse samconfig.toml: {}", e)))?;

        // Try deploy.parameters first, then global.parameters
        let paths = [&["deploy", "parameters"][..], &["global", "parameters"][..]];
        for path in &paths {
            let mut section = data.get(config_env);
            for key in *path {
                section = section.and_then(|s| s.get(key));
            }
            if let Some(name) = section.and_then(|s| s.get("stack_name")) {
                if let Some(s) = name.as_str() {
                    return Ok(s.trim_matches('"').to_string());
                }
            }
        }

        Err(CliError::Config(format!(
            "no stack_name found in samconfig.toml for config-env '{}'",
            config_env
        )))
    }

    /// Query CloudFormation for the deployed API endpoint URL.
    pub fn stack_endpoint(&self, r: &dyn Runner, config_env: &str) -> Result<String, CliError> {
        if self.dry_run {
            return Ok(format!(
                "https://DRY-RUN.execute-api.us-east-1.amazonaws.com/{}",
                config_env
            ));
        }

        let stack_name = self.read_stack_name(config_env)?;
        let result = r.run(
            &[
                "aws",
                "cloudformation",
                "describe-stacks",
                "--stack-name",
                &stack_name,
                "--query",
                "Stacks[0].Outputs[?OutputKey=='SupervictorApiEndpoint'].OutputValue",
                "--output",
                "text",
            ],
            &RunOptions {
                env: self.env.clone(),
                verbose: self.verbose,
                capture: true,
                ..Default::default()
            },
        )?;

        let url = result.stdout.trim().trim_end_matches('/').to_string();
        if url.is_empty() {
            return Err(CliError::Config(format!(
                "no SupervictorApiEndpoint output found for stack '{}'",
                stack_name
            )));
        }
        Ok(url)
    }

    /// Run sam deploy. Returns true if changes were deployed.
    pub fn deploy(
        &self,
        r: &dyn Runner,
        config_env: &str,
        force_upload: bool,
    ) -> Result<bool, CliError> {
        let log_path = self
            .config
            .log_dir
            .join(format!("sam_deploy_{}.log", config_env));
        runner::step(&format!("Deploying to {} stack", config_env));

        let mut cmd_parts: Vec<String> = vec![
            "sam".into(),
            "deploy".into(),
            "--config-env".into(),
            config_env.into(),
        ];

        // Override samconfig values from env vars
        let env = self.env.as_ref();
        if let Some(v) = env
            .and_then(|e| e.get("SAM_STACK_NAME"))
            .filter(|v| !v.is_empty())
        {
            cmd_parts.extend(["--stack-name".into(), v.clone()]);
        }
        if let Some(v) = env
            .and_then(|e| e.get("SAM_REGION"))
            .filter(|v| !v.is_empty())
        {
            cmd_parts.extend(["--region".into(), v.clone()]);
        }
        if let Some(v) = env
            .and_then(|e| e.get("SAM_S3_PREFIX"))
            .filter(|v| !v.is_empty())
        {
            cmd_parts.extend(["--s3-prefix".into(), v.clone()]);
        }

        // Build --parameter-overrides
        let param_map = [
            ("SAM_ENVIRONMENT", "Environment"),
            ("SAM_APP_NAME", "AppName"),
            ("SAM_STACK_NAME", "StackName"),
            ("SAM_TRUSTSTORE_URI", "TruststoreUri"),
        ];
        let mut param_parts: Vec<String> = Vec::new();
        for (env_key, cfn_param) in &param_map {
            if let Some(val) = env.and_then(|e| e.get(*env_key)).filter(|v| !v.is_empty()) {
                param_parts.push(format!("{}={}", cfn_param, val));
            }
        }
        if !param_parts.is_empty() {
            cmd_parts.push("--parameter-overrides".into());
            cmd_parts.push(param_parts.join(" "));
        }

        if force_upload {
            cmd_parts.push("--force-upload".into());
        }

        let cmd_refs: Vec<&str> = cmd_parts.iter().map(|s| s.as_str()).collect();
        let result = r.run(
            &cmd_refs,
            &RunOptions {
                cwd: Some(self.config.cloud_dir.clone()),
                env: self.env.clone(),
                verbose: self.verbose,
                dry_run: self.dry_run,
                check: false,
                log_to: Some(log_path.clone()),
                ..Default::default()
            },
        )?;

        if result.status != 0 {
            let combined = format!("{}{}", result.stdout, result.stderr);
            if combined.contains("No changes to deploy") {
                runner::success("Stack is already up to date.");
                return Ok(false);
            }
            runner::error(&format!("sam deploy failed (see {})", log_path.display()));
            return Err(CliError::Command {
                cmd: cmd_parts.join(" "),
                code: result.status,
                stderr: result.stderr,
            });
        }

        runner::success(&format!("Deployed to {}", config_env));
        Ok(true)
    }

    fn write_env_overrides(&self) -> Result<PathBuf, CliError> {
        let dir = std::env::temp_dir();
        let path = dir.join(format!("sam_env_{}.json", std::process::id()));
        fs::write(&path, LOCAL_ENV_OVERRIDES)?;
        Ok(path)
    }
}

#[cfg(test)]
#[path = "sam_tests.rs"]
mod tests;
